Improve String#demodulize method

alfonsojimenez opened the pull request:
### Benchmark

Using: **ruby 2.3.0**

``` ruby
# frozen_string_literal: true

require 'benchmark'

class Inflector
  def self.original_demodulize(path)
    path = path.to_s
    if i = path.rindex('::')
      path[(i+2)..-1]
    else
      path
    end
  end

  def self.new_demodulize(path)
    (i = path.rindex('::')) ? path[i + 2..-1] : path
  end
end

n = 1_000_000
path = 'String::Inflections'

Benchmark.bmbm do |x|
  x.report('.original_demodulize') do
    n.times { Inflector.original_demodulize(path) }
  end

  x.report('.new_demodulize') do
    n.times { Inflector.new_demodulize(path)  }
  end
end
```
#### Worst case

When `path` contains `::` (e.g: `'String::Inflections'`)

```
Rehearsal --------------------------------------------------------
#original_demodulize   0.320000   0.000000   0.320000 (  0.326918)
#new_demodulize        0.300000   0.000000   0.300000 (  0.295983)
----------------------------------------------- total: 0.620000sec

                           user     system      total        real
#original_demodulize   0.320000   0.000000   0.320000 (  0.321361)
#new_demodulize        0.270000   0.000000   0.270000 (  0.271659)
```

**Improvement**: _~15.62%_
#### Best case

When `path` does not contain `::` (e.g: `''`)

```
Rehearsal --------------------------------------------------------
#original_demodulize   0.140000   0.000000   0.140000 (  0.145417)
#new_demodulize        0.120000   0.000000   0.120000 (  0.115560)
----------------------------------------------- total: 0.260000sec

                           user     system      total        real
#original_demodulize   0.150000   0.000000   0.150000 (  0.145593)
#new_demodulize        0.110000   0.000000   0.110000 (  0.113746)
```

**Improvement**: _~26.67%_

alfonsojimenez committed:
Improve String#demodulize method

rafaelfranca commented:
What is the improvement here? I only see cosmetic changes in the diff.

alfonsojimenez commented:
@rafaelfranca Unnecessary ```path.to_s``` and freezing ```::```

rafaelfranca commented:
`path.to_s` is no unnecessary. It is to make it working when calling `demodulize` on a symbol. And `freeze` we don't want to introduce in the codebase.

Could you add a test to make sure `demodulize` work with symbols?

alfonsojimenez commented:
Sorry @rafaelfranca, I can't think of a case when ```path``` can be a symbol. Also, I see several frozen '::' strings within the same class

alfonsojimenez commented:
@rafaelfranca I have removed the ```#freeze``` and added the ```frozen_string_literal``` pragma. The performance improvement is still the same for ruby 2.3.0. It also implicitly freeze some other strings (actually just a few methods contain not frozen strings in this file).

Regarding to ```path.to_s```, could you please provide an example when ```path``` is a symbol?

rafaelfranca commented:
We still don't have any example, but if you look the commit that added that line it added tests to assert some methods in this file work with symbol. So `Inflector.demodulize(:'Foo::Bar')` should do it.

jeremy commented on activesupport/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+*   Improve `String#demodulize` method
```
No need for changelog entry for a perf tweak since it doesn't affect users.

jeremy commented on activesupport/lib/active_support/inflector/methods.rb:
```diff
@@ -195,12 +197,7 @@ def dasherize(underscored_word)
-      else
-        path
-      end
+      (i = path.rindex('::')) ? path[i + 2..-1] : path
```
Ternary is harder to visually parse. `i + 2..-1` spacing makes it look like we're adding 1 to a Range.

jeremy commented on activesupport/lib/active_support/inflector/methods.rb:
```diff
@@ -195,12 +197,7 @@ def dasherize(underscored_word)
     #
     # See also #deconstantize.
     def demodulize(path)
-      path = path.to_s
```
`path` may be a non-String that needs explicit coercion before demodulizing.

alfonsojimenez replied:
@jeremy Thanks for your comments. I've rebased the branch.

Must this method be responsible for the string coercion? After these changes this will not work anymore:

`ActiveSupport::Inflector.demodulize(:'Foo::Bar')`

or

`ActiveSupport::Inflector.demodulize(BigDecimal.new('1.00'))`

But I can't think of a use case for these examples. IHMO, it makes more sense to leave that responsibility outside this method.

cc @rafaelfranca

jeremy replied:
It does make sense for non-string-like objects, but it also breaks
compatibility. Is anyone calling this with symbols, for example?
On Mon, May 2, 2016 at 08:00 Alfonso Jiménez <notifications@github.com>
wrote:

> In activesupport/lib/active_support/inflector/methods.rb
> <https://github.com/rails/rails/pull/23985#discussion_r61750121>:
>
> > @@ -195,12 +197,7 @@ def dasherize(underscored_word)
> >      #
> >      # See also #deconstantize.
> >      def demodulize(path)
> > -      path = path.to_s
>
> @jeremy <https://github.com/jeremy> Thanks for your comments. I've
> rebased the branch.
>
> Must this method be responsible for the string coercion? After these
> changes this will not work anymore:
>
> ActiveSupport::Inflector.demodulize(:'Foo::Bar')
>
> or
>
> ActiveSupport::Inflector.demodulize(BigDecimal.new('1.00'))
>
> But I can't think of a use case for these examples. IHMO, it makes more
> sense to leave that responsibility outside this method.
>
> cc @rafaelfranca <https://github.com/rafaelfranca>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/23985/files/0fbc14189814e1e48dc48c391e7a3191316f5839#r61750121>
>

rafaelfranca commented:
Seems like we wanted to deprecate usage of symbols before but we forget to remove the code that make it support before removing the deprecation https://github.com/rails/rails/commit/3718ccd2a61c2c189913bcfd487912f592fa0660.

If we are going to remove symbol support we should deprecate it again.

alfonsojimenez committed:
Bringing back deprecation on ActiveSupport::Inflector.demodulize

alfonsojimenez commented:
@rafaelfranca @jeremy Ah right :smile: I've deprecated the usage of symbols on ```ActiveSupport::Inflector.demodulize``` again (based on https://github.com/rails/rails/commit/3718ccd2a61c2c189913bcfd487912f592fa0660)
